streaming/base/dataset/streaming.py

```diff
@@ -0,0 +1,623 @@
+        self.total = len(sample_ids)
+        self.iter_index = 0
+        self.download_index = 0
+        self.is_stopped = False
+
+    def stop(self) -> None:
+        """Stop the thread and exit."""
+        self.is_stopped = True
+
+    def __iter__(self) -> Iterator[int]:
+        """Iterate over our samples while waiting for them to download first.
+
+        Returns:
+            Iterator[int]: Each sample, having been downloaded.
+        """
+        while self.iter_index < self.total:
+            if self.iter_index < self.download_index:
+                yield self.sample_ids[self.iter_index]
+                self.iter_index += 1
+                continue
+            if self.is_stopped:
+                break
+            sleep(TICK)
+
+
+class Dataset(IterableDataset):
+    """A streaming pytorch IterableDataset that is also resumable mid-epoch.
+
+    Checkpoints are represented in JSON as follows:
+
+        {
+            'epoch': int,
+            'sample_in_epoch': int,
+        }
+
+    Args:
+        local (str): Local dataset directory where shards are cached by split.
+        remote (str, optional): Download shards from this remote path or directory. If None, this
+            rank and worker's partition of the dataset must all exist locally. Defaults to
+            ``None``.
+        split (str, optional): Which dataset split to use, if any. Defaults to ``None``.
+        shuffle (bool): Whether to iterate over the samples in randomized order. Defaults to
+            ``False``.
+        predownload (int, optional): Target number of samples ahead to download the shards of while
+            iterating. Defaults to ``100_000``.
+        keep_zip (bool, optional): Whether to keep or delete the compressed file when
+            decompressing downloaded shards. If set to None, keep iff remote is local. Defaults to
+            ``None``.
+        download_retry (int): Number of download re-attempts before giving up. Defaults to ``2``.
+        download_timeout (float): Number of seconds to wait for a shard to download before raising
+            an exception. Defaults to ``60``.
+        validate_hash (str, optional): Optional hash or checksum algorithm to use to validate
+            shards. Defaults to ``None``.
+        shuffle_seed (int, optional): Seed for shuffling, or ``None`` for random seed. Defaults to
+            ``None``.
+        shuffle_world_size (int, optional): Canonical world size for shuffling. Defaults to
+            ``None``, which is interpreted as the world size of the initial run.
+        batch_size (int, optional): Batch size of its DataLoader, which affects how the dataset is
+            partitioned over the workers. Defaults to ``None``.
+    """
+
+    def __init__(self,
+                 local: str,
+                 remote: Optional[str] = None,
+                 split: Optional[str] = None,
+                 shuffle: bool = False,
+                 predownload: Optional[int] = 100_000,
```

Comment:
is this number `100_000` realistic? Because this number is for every `num_workers` process right? For example, let say, you are doing a model training on a 1 node with 8 GPUs with 8 `num_workers` each per GPU, which means total 64 num_process per node, which leads to 64 * 100_000 = 6_400_000 number of samples !!! 

Should we lower the default number of `10_000` maybe !! Whats your thought on that! Also, leaving as it is causes no harm at all. Wanted to know your thoughts on choosing `100_000` !! Thanks!

Reply:
about 60k samples/shard in c4

Reply:
100K per worker. no, workers do not generate shard files